### PR TITLE
ADD: Document labeling

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/target_models/document_labeling.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+def get_context(item, dpi_norm):
+    context = []
+    for page in item:
+        for token in page["tokens"]:
+            pos = token["position"]
+            offset = token["doc_offset"]
+            if dpi_norm:
+                dpi = page["pages"][0]["dpi"]
+#                dpi = item["pages"][token["page_num"]]["dpi"]
```
Comment (benleetownsend) [inline finetune/target_models/document_labeling.py]: Oops

File: finetune/target_models/document_labeling.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                x_norm = 300 / dpi["dpix"]
+                y_norm = 300 / dpi["dpiy"]
+            else:
+                x_norm = 1.
+                y_norm = 1.
+
+            context.append(
+                {
+                    'top': pos["top"] * y_norm,
```
Comment (benleetownsend) [inline finetune/target_models/document_labeling.py:19]: Strange indent

File: finetune/target_models/document_labeling.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
+from finetune.target_models.sequence_labeling import SequenceLabeler
+
+def get_context(item, dpi_norm):
```
Comment (benleetownsend) [inline finetune/target_models/document_labeling.py]: Item should be renamed to document or something

File: finetune/target_models/document_labeling.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            input_level=input_level
+        )
+        text.extend(t)
+        context.extend(c)
+        labels_out.extend(l)
+        offsets.extend(o)
+    return text, context, labels_out, offsets
+        
+
+class DocumentLabeler(SequenceLabeler):
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py:129]: Maybe it's worth adding a quick comment to specify which version of pdfextraction output this takes in?

File: finetune/target_models/document_labeling.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    return output
+
+def _single_convert_to_finetune(document, labels=None, dpi_norm=True, input_level="page", doc_id=None):
+    context = get_context(document, dpi_norm)
+    texts = []
+    offsets = []
+    
+    if input_level.lower() == "page":
+        for page in document:
+            page_obj = page["pages"][0]
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py:61]: I'm a bit confused. pdfextraction output looks like a document is a dict with a "pages" attribute which is a list of pages. In this function, though, it looks like we're assuming that a document is a list of pages, where each page's "pages" attribute is a single element list.

File: finetune/target_models/document_labeling.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            texts.append(page_obj["text"])
+            offsets.append(page_obj["doc_offset"])
+    elif input_level.lower() == "block":
+        for page in document:
+            for block_obj in page["blocks"]:
+                texts.append(block_obj["text"])
+                offsets.append(block_obj["doc_offset"])
+    else:
+        raise ValueError("input_level should be either page or block")
+    if doc_id is not None:
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py]: this is to give us an option to overwrite the doc_id?

File: finetune/target_models/document_labeling.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    output = [[] for _ in offsets]
+    page_idx = 0
+    page_start = offsets[0]["start"]
+    page_end = offsets[0]["end"]
+    for block in to_reblock:
+        block = dict(block)
+        while not (page_start <= block["start"] < page_end):
+            page_idx += 1
+            if page_idx == len(offsets):
+                raise ValueError("Block: {} does not not align with offsets")
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py]: missing .format()

File: finetune/target_models/document_labeling.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		    'start': offset["start"],
+		    'end': offset["end"],
+                }
+	    )
+    return context
+
+def reblock_by_offsets(offsets, texts, to_reblock):
+    if len(offsets) != len(texts):
+        raise ValueError("Length of offsets and texts do not match")
+    to_reblock = sorted(to_reblock, key=lambda x: x["start"])
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py:33]: does "start" actually hold across pages?

File: finetune/target_models/document_labeling.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                    'left': pos["left"] * x_norm,
+		    'right': pos["right"] * x_norm,
+		    'text': token["text"],
+		    'start': offset["start"],
+		    'end': offset["end"],
+                }
+	    )
+    return context
+
+def reblock_by_offsets(offsets, texts, to_reblock):
```
Comment (lhz1029) [inline finetune/target_models/document_labeling.py:30]: in what case does the output that comes out of odl need to be reblocked? I get where we might want modify the indices if they don't start at 0, but how about the reblocking?

File: tests/test_document.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                {'start': 2023, 'end': 2040, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 2450, 'end': 2467, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 2963, 'end': 2980, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 4713, 'end': 4730, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 5445, 'end': 5462, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 6642, 'end': 6659, 'label': 'city of hollywood', 'text': 'City of Hollywood'},
+                {'start': 7082, 'end': 7099, 'label': 'city of hollywood', 'text': 'City of Hollywood'}
+            ]
+        ]
+        with open("tests/data/test_ocr_documents.jl", "rt") as fp:
```
Comment (madisonmay) [inline tests/test_document.py]: weird to have jl file that is json

File: finetune/target_models/document_labeling.py (hunk 9 above)
Comment (benleetownsend) [reply finetune/target_models/document_labeling.py:30]: odl output is at document level and we need to take in input at page level.

File: finetune/target_models/document_labeling.py (hunk 8 above)
Comment (benleetownsend) [reply finetune/target_models/document_labeling.py:33]: the labels are at a document level.

File: finetune/target_models/document_labeling.py (hunk 5 above)
Comment (benleetownsend) [reply finetune/target_models/document_labeling.py:61]: This is what comes back from pdfextraction when top level is set to page. A list of dicts each containing a single page.

File: finetune/target_models/document_labeling.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        offsets.extend(o)
+        if l is not None:
+            labels_out.extend(l)
+    return text, context, labels_out, offsets
+        
+
+class DocumentLabeler(SequenceLabeler):
+    """
+    A wrapper to use SequenceLabeler ontop of indico's PDFExtraction APi 
+    in ondocument mode with labels at a document charachter level.
```
Comment (benleetownsend) [inline finetune/target_models/document_labeling.py:132]: Spelling

**Pull request closed without merging**